refactor(core): add links to top 10 error messages

twerske opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
No links to angular.io/errors guides

Issue Number: N/A


## What is the new behavior?
Links to the new guides!

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -191,6 +191,12 @@ export class NgCompiler {
 
+    this.diagnostics.forEach(diag => {
+      if (diag.code && HAS_GUIDE.has(ngErrorCode(diag.code))) {
+        diag.messageText += `. Find more at https://angular.io/errors/NG${ngErrorCode(diag.code)}`;
```
I'd propose extracting error page base URL to a separate const (for both runtime and compiler files), so it's easier to manage it in case the base path should be different inside g3.

```suggestion
// Base URL for the error details page.
// Keep this value in sync with a similar const in `packages/core/src/render3/error_code.ts`.
const ERROR_DETAILS_PAGE_BASE_URL = 'https://angular.io/errors';
...
        diag.messageText += `. Find more at ${ERROR_DETAILS_PAGE_BASE_URL}/NG${ngErrorCode(diag.code)}`;
```

AndrewKushnir commented on packages/core/src/render3/error_code.ts:
```diff
@@ -41,5 +41,15 @@ export class RuntimeError extends Error {
 export function formatRuntimeError(code: RuntimeErrorCode, message: string): string {
   const fullCode = code ? `NG0${code}: ` : '';
+
+  const HAS_GUIDE = new Set([
```
We should move this const outside of the `formatRuntimeError` function, so that it's created once (vs creating this Set each time we invoke this function).

twerske replied:
Agreed! I was running into a lint issue: `no-toplevel-property-access  Avoid toplevel property access.` with the const defined outside the function - is there a valid fix besides a lint rule exception?

AndrewKushnir commented on packages/core/src/render3/error_code.ts:
```diff
@@ -41,5 +41,15 @@ export class RuntimeError extends Error {
+    RuntimeErrorCode.EXPORT_NOT_FOUND
+  ]);
+
+  if (HAS_GUIDE.has(code)) {
+    return `${fullCode}${message}. Find more at https://angular.io/errors/NG0${code}`;
+  }
   return `${fullCode}${message}`;
```
```suggestion
  let errorMessage = `${fullCode}${message}`;
  if (HAS_GUIDE.has(code)) {
    errorMessage = `${errorMessage}. Find more at https://angular.io/errors/NG0${code}`;
  }
  return errorMessage;
```

AndrewKushnir commented on packages/core/src/render3/error_code.ts:
```diff
@@ -41,5 +41,15 @@ export class RuntimeError extends Error {
 export function formatRuntimeError(code: RuntimeErrorCode, message: string): string {
   const fullCode = code ? `NG0${code}: ` : '';
+
+  const HAS_GUIDE = new Set([
```
May be also add a quick comment above the `HAS_GUIDE` const to describe what it is:
```suggestion
  // Contains a set of error messages that have details guides at angular.io.
  // Full list of available error guides can be found at https://angular.io/...
  const HAS_GUIDE = new Set([
```

AndrewKushnir reviewed (changes requested):
The change looks good, just left a few comments. Thank you.

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -191,6 +191,17 @@ export class NgCompiler {
+        // Base URL for the error details page.
+        // Keep this value in sync with a similar const in
+        // `packages/core/src/render3/error_code.ts`.
+        const ERROR_DETAILS_PAGE_BASE_URL = 'https://angular.io/errors';
```
I think it'd be great to move this const to the `packages/compiler-cli/src/ngtsc/diagnostics/src/error_code.ts` alongside the `HAS_GUIDE` const, so that the structure is similar to what we have at runtime (in `packages/core/src/render3/error_code.ts`).

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/diagnostics/src/error_code.ts:
```diff
@@ -161,6 +161,15 @@ export enum ErrorCode {
+export const HAS_GUIDE = new Set([
+  ErrorCode.DECORATOR_ARG_NOT_LITERAL, ErrorCode.PARAM_MISSING_TOKEN,
+  ErrorCode.SCHEMA_INVALID_ELEMENT, ErrorCode.SCHEMA_INVALID_ATTRIBUTE,
+  ErrorCode.MISSING_REFERENCE_TARGET
```
Minor comment: may be add a comma at the end of last line - the code should be auto re-formatted to have each error code on its own line (same for `packages/core/src/render3/error_code.ts` file).
```suggestion
  ErrorCode.MISSING_REFERENCE_TARGET,
```

twerske replied:
That was my thought but linting did the opposite (converted them to individual lines) - I can override though, maybe my personal settings are configured incorrectly

AndrewKushnir replied:
The linting behavior looks correct, but adding an extra comma after the last element typically forces linter to put values on their own lines :)

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/diagnostics/src/error_code.ts:
```diff
@@ -161,6 +161,15 @@ export enum ErrorCode {
+/**
+ * @internal
+ */
+export const HAS_GUIDE = new Set([
```
@twerske for some reasons I can't reply to your previous comment about "no-toplevel-property-access" tslint rule, so I'll post a reply here. Since the `RuntimeErrorCode` is a const enum, its values are inlined in the generated code ([see an example here](https://www.typescriptlang.org/play?#code/MYewdgzgLgBApmArgWxgJUWKBLZcCiAToSIQMIgAmcMA3gFAwwD0zMZAFgIZgDmNAEThQ4wHOBhEShCI0kANAApp8AZVUBJAPIA5APpkAEgEEdAcXwC9xgGIAVfGgOH8ZANKWYAXhgByAIwADIG+ADT0cqwwQgAOCNRgwACeMBpgAFai4mCSxKSyTGQAmmQAMhpkegIaVfiK+DoCDcXefgBMwWFyyloAahpNTjpadno2WgCqja2+Hf5dAL4RoJCwcMStGFi4BHnkVHAAdMVlFVU1TfWNzUUA3EA)), so the lint rule can be safely ignored, i.e. you can disable that tslint rule for the entire `HAS_GUIDE` block.

```suggestion
/* tslint:disable:no-toplevel-property-access */
export const HAS_GUIDE = new Set([ ... ]);
/* tslint:enable:no-toplevel-property-access */
```

AndrewKushnir commented on packages/core/src/render3/error_code.ts:
```diff
@@ -38,8 +38,25 @@ export class RuntimeError extends Error {
+    // Base URL for the error details page.
+    // Keep this value in sync with a similar const in
+    // `packages/compiler-cli/src/ngtsc/core/src/compiler.ts`.
+    const ERROR_DETAILS_PAGE_BASE_URL = 'https://angular.io/errors';
```
I'd propose moving this const at the top level (closer to the top of that file, above the `RuntimeErrorCode` enum).

AndrewKushnir commented on packages/core/src/render3/error_code.ts:
```diff
@@ -38,8 +38,25 @@ export class RuntimeError extends Error {
-  return `${fullCode}${message}`;
+
+  let errorMessage = `${fullCode}${message}`;
+  if (HAS_GUIDE.has(code)) {
```
I'd like to propose a slightly different name for the `HAS_GUIDE` const so that this condition looks a bit nicer :)
```suggestion
  if (ERRORS_WITH_GUIDES.has(code)) {
```

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/diagnostics/src/error_code.ts:
```diff
@@ -161,6 +161,25 @@ export enum ErrorCode {
+export const ERROR_DETAILS_PAGE_BASE_URL = 'https://angular.io/errors';
+
+/**
+ * @internal
```
I'd propose adding the same comment that was added for similar const in runtime code:

```
/**
 * @internal
 * Contains a set of error messages that have details guides at angular.io.
 * Full list of available error guides can be found at https://angular.io/errors
 */
```

AndrewKushnir commented on packages/core/src/render3/error_code.ts:
```diff
@@ -38,8 +43,25 @@ export class RuntimeError extends Error {
 
+// Contains a set of error messages that have details guides at angular.io.
+// Full list of available error guides can be found at https://angular.io/errors
+/* tslint:disable:no-toplevel-property-access */
```
Could you please add a quick comment on why it's safe to disable this tslint rule? You can find an example here: https://github.com/angular/angular/blob/master/packages/core/src/di/metadata.ts#L106 (you can reuse the same comment and just update the enum name).

AndrewKushnir reviewed (approved):
Two minor comments, otherwise LGTM 👍 Thanks @twerske.

AndrewKushnir commented:
@twerske this PR would also need a rebase, since there is a potential merge conflict with the master branch. Thank you.

zarend commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -188,10 +189,26 @@ export class NgCompiler {
    */
   getDiagnosticsForFile(file: ts.SourceFile, optimizeFor: OptimizeFor): ts.Diagnostic[] {
-    return [
+    return this.addMessageTextDetails([
```
👍

zarend commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -179,7 +179,8 @@ export class NgCompiler {
    */
   getDiagnostics(): ts.Diagnostic[] {
-    return [...this.getNonTemplateDiagnostics(), ...this.getTemplateDiagnostics()];
+    return this.addMessageTextDetails(
```
👍

zarend commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -188,10 +189,26 @@ export class NgCompiler {
+  /**
+   * Add Angular.io error guide links to diagnostics for this compilation.
+   */
+  addMessageTextDetails(diagnostics: ts.Diagnostic[]): ts.Diagnostic[] {
```
Make this private because it's not part of `NgCompiler`'s CLI.

zarend commented on packages/compiler-cli/integrationtest/bazel/injector_def/ivy_build/app/test/module_spec.ts:
```diff
@@ -63,9 +63,11 @@ describe('Ivy NgModule', () => {
       }
 
       const errorCode = ivyEnabled ? 'NG0200: ' : '';
+      const errorLink = ivyEnabled ? '. Find more at https://angular.io/errors/NG0200' : '';
```
This is not a blocker for me, but I don't like that we have conditional logic of what the expected behavior of a test is. That's because it's not clear which code flow we are actually testing. Are we testing the ivyEnabled case or legacy view-engine case? I don't know what kind of coverage to expect from this test.

It would be better if we made sure `ivyEnabled` was true for this test, or broke it up into two separate tests.

Another thing that would be better is if we expected it to throw an error but did not check the error message, or only tested that the error message contained the substring  `Circular dependency in DI detected for AModule. Dependency path: AModule > BModule > AModule`

twerske replied:
Most tests go with the final option you propose (testing the substring), though it'd be nice to have some tests check the code and link formatting. These tests currently cover both `ivyEnabled = true | false` which isn't great and likely should be refactored but is probably outside scope of this PR.

AndrewKushnir replied:
@zarend we use both approaches (one test with `ivyEnabled` flag inside and separate tests for VE/Ivy) in different tests. If the delta between VE and Ivy behavior is minimal (like in this case), I find having a flag more convenient. Eventually all these flags go away once VE is removed.

zarend replied:
fair enough, thanks for adding some context. I'm okay with it being like this :)

zarend reviewed (changes requested):
Please make `addMessageTextDetails` a private method. We want to be careful about which things we expose on `NgCompiler` :)

zarend reviewed (approved):
👍

AndrewKushnir commented on packages/core/test/bundling/todo/bundle.golden_symbols.json:
```diff
@@ -92,6 +92,9 @@
   {
     "name": "R3ViewContainerRef"
   },
+  {
+    "name": "RUNTIME_ERRORS_WITH_GUIDES"
+  },
```
@twerske sorry I had to remove this PR from the merge queue due to this change. Ideally, we should not see the `RUNTIME_ERRORS_WITH_GUIDES` symbol in golden files, since they are supposed to represent dev-mode-only errors, which should be tree-shaken away at build step (thus should not appear in these files). The problem is that the set would be included into an app bundle in prod mode and as the set grows, there will be more unused code pulled in. We should find out why the `RUNTIME_ERRORS_WITH_GUIDES` symbol get retained and try to resolve that (most likely one of the errors is thrown without the `ngDevMode` guard).

twerske committed:
refactor(core): add links to top runtime errors

add links to 5 runtime error messages
navigate user to AIO new /errors pages for debugging

twerske committed:
refactor(core): add links to top compiler errors

add links to 5 compiler error messages
navigate user to AIO new /errors pages for debugging

AndrewKushnir reviewed (approved):
@twerske thanks for applying additional changes 👍

AndrewKushnir commented:
Merge-assistance: this PR would require [this g3 patch](http://cl/351911975) internally.

AndrewKushnir commented:
Just a quick note: I've added a blocked label for now since landing this PR requires internal g3 patch (which should be applied before merging the change) and we'll do that on Tuesday next week (Jan 19).
